projects/ADuCM360_demo_cn0411/src/CN0411.c

```diff
@@ -0,0 +1,1876 @@
+ * @param cn0411_dev - The device structure.
+ * @return none
+*/
+void CN0411_cmd_tds(uint8_t *args, cn0411_device *cn0411_dev)
+{
+	int32_t ret;
+
+	printf("TDS measurement in progress...\n");
+	ret = CN0411_compute_tds(cn0411_dev);
+	if(ret == CN0411_SUCCESS) {
+		printf("TDS = %.10f\n", cn0411_dev->tds);
+	} else {
+		printf("Get TDS value failed!\n");
+	}
+}
+
+/**
+ * Internal interrupt handler for UART
+
+ * @return none
+*/
+void CN0411_interrupt(void)
+{
+	unsigned short status;
+	char c;
+
+	status = UrtIntSta(pADI_UART);
+	if (status & COMIIR_NINT) {
+		return;   /* Check if UART is busy */
+	}
+	switch (status & COMIIR_STA_MSK) {
+	case COMIIR_STA_RXBUFFULL:
+		UART_ReadChar(&c);
+		switch(c) {
+		case _BS:
+			if (uart_rcnt) {
+				uart_rcnt--;
+				uart_rx_buffer[uart_rcnt] = 0;
+				UART_WriteChar(c, UART_WRITE_IN_INT);
+				UART_WriteChar(' ', UART_WRITE_IN_INT);
+				UART_WriteChar(c, UART_WRITE_IN_INT);
+			}
+			break;
+		case _CR: /* Check if read character is ENTER */
+			uart_cmd = UART_TRUE;                    /* Set flag */
+			break;
+		default:
+			uart_rx_buffer[uart_rcnt++] = c;
+
+			if (uart_rcnt == UART_RX_BUFFER_SIZE) {
+				uart_rcnt--;
+				UART_WriteChar(_BS, UART_WRITE_IN_INT);
+			}
+			UART_WriteChar(c, UART_WRITE_IN_INT);
+		}
+		uart_rx_buffer[uart_rcnt] = '\0';
+		break;
+	case COMIIR_STA_TXBUFEMPTY:
+		if (uart_tcnt) {
+			uart_tbusy = UART_TRUE;
+			uart_tcnt--;
+			UART_WriteChar(uart_tx_buffer[uart_tpos++], UART_WRITE);
+			if (uart_tpos == UART_TX_BUFFER_SIZE) {
+				uart_tpos = 0;
+			}
+		} else {
+			uart_tbusy = UART_FALSE;
+		}
+		break;
+	default:
+		;
+	}
+}
+
+/**
+ * Set PWM frequency
+ *
+ * @param freq - frequency value to be set.
+ * @return none
+*/
+void CN0411_pwm_freq(uint16_t freq)
+{
+	switch (freq) {
+	case PWM_FREQ_94:
+		pwm_freq = pwm_100_freq;
+		pwm_tick_count = 0;
+		pwm_index = 0;
+		break;
+	case PWM_FREQ_2400:
+		pwm_freq = pwm_2400_freq;
+		pwm_tick_count = 0;
+		pwm_index = 0;
+		break;
+	default:
+		break;
+	}
+}
+
+/**
+ * Generate PWM
+
+ * @return none
+*/
+void CN0411_pwm_gen(void)
+{
+	switch(pwm_status) {
+	case PWM_SYSCALIB_AIN7:
+		DioClr(pADI_GP1, BIT4);
+		DioSet(pADI_GP1, BIT2|BIT3);
+		break;
+	case PWM_SYSCALIB_AIN8:
+		DioClr(pADI_GP1, BIT2|BIT3);
+		DioSet(pADI_GP1, BIT4);
+		break;
+	case PWM_CONVERSION:
+		if(pwm_tick_count < pwm_freq[pwm_index]) {
+			pwm_tick_count++;
+		} else {
+			pSetClr[pwm_setclr[pwm_index]](pADI_GP1, pwm_bit[pwm_index]);
+			if(pwm_index >= ARRAY_SIZE(pwm_100_freq)-1) {
+				pwm_index = 0;
+				pwm_tick_count = 0;
+			} else {
+				pwm_index++;
+				pwm_tick_count++;
+			}
+		}
+		break;
+	default:
+		break;
+	}
+}
+
+/**
+ * CN0411 Initialization
+ *
+ * @param cn0411_dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+*/
+int32_t CN0411_init(cn0411_device *cn0411_dev)
+{
+	int32_t ret;
+
+	/* Initialize Device Structure */
+	cn0411_dev->v_dac = DAC_OUT_DEFAULT_VAL;
```

Comment:
you could use a init structure and initialize it in the declaration section. It would make configuring cn0411 a lot easier, no functions would require modification to change configurations.